drivers/docker/driver.go

```diff
@@ -536,10 +536,30 @@ func (d *Driver) loadImage(task *drivers.TaskConfig, driverConfig *TaskConfig, c
 
 func (d *Driver) containerBinds(task *drivers.TaskConfig, driverConfig *TaskConfig) ([]string, error) {
 
-	allocDirBind := fmt.Sprintf("%s:%s", task.TaskDir().SharedAllocDir, task.Env[env.AllocDir])
-	taskLocalBind := fmt.Sprintf("%s:%s", task.TaskDir().LocalDir, task.Env[env.TaskLocalDir])
-	secretDirBind := fmt.Sprintf("%s:%s", task.TaskDir().SecretsDir, task.Env[env.SecretsDir])
-	binds := []string{allocDirBind, taskLocalBind, secretDirBind}
+	binds := []string{
+		// alloc directory
+		fmt.Sprintf("%s:%s", task.TaskDir().SharedAllocDir, task.Env[env.AllocDir]),
+		// task local directory
+		fmt.Sprintf("%s:%s", task.TaskDir().LocalDir, task.Env[env.TaskLocalDir]),
+		// secrets dir bind
+		fmt.Sprintf("%s:%s", task.TaskDir().SecretsDir, task.Env[env.SecretsDir]),
+	}
+
+	for _, m := range task.Mounts {
+		opts := ""
+		if m.Readonly {
+			opts = ":ro"
+		}
+		binds = append(binds, fmt.Sprintf("%s:%s%s", m.HostPath, m.TaskPath, opts))
+	}
+
+	for _, m := range task.Devices {
+		opts := ":ro"
+		if m.Permissions == "rw" {
+			opts = ""
+		}
+		binds = append(binds, fmt.Sprintf("%s:%s%s", m.HostPath, m.TaskPath, opts))
```

Comment:
Surprisingly the `task.Devices` and `task.Mounts` are extremely similar and are to be handled the same way, both being bind-mount based - we should consider using the merging them.

I would have expected the Devices to be Mknod specifications - or had Volumes be flexible to allow for docker plugin volumes (which understandably is docker specific).

I'll follow up on this outside this PR.

Reply:
Umm.. Thanks!  I will switch to using Docker Driver API, since I guess our Device struct is based on it.

My impression is that the underlying containerization technologies treat devices and mounts very differently - though the Docker API high level Devices and Mounts is effectively the same as you can do all API device mounting via mounting.  However, libcontainer[1] and LXC/LXD[2] allows for setting Mknod arguments so one can theoretically make a device available on a container without making it available on host `/dev`.

[1] https://godoc.org/github.com/opencontainers/runc/libcontainer/configs#Device
[2] https://github.com/lxc/lxd/blob/master/doc/containers.md#devices-configuration